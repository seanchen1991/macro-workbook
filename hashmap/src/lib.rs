#[macro_export]
macro_rules! hashmap {
    // matches one or more key-value pairs that end with a comma
    // this rule is necessary in order to match single key-value
    // pairs that end with a trailing comma
    ($($key:expr => $value:expr,)+) => { $crate::hashmap!($($key => $value),+) };
    // matches zero or more key-value pairs that are each 
    // separated by commas
    ($($key:expr => $value:expr),*) => {
        {
            let mut _map = ::std::collections::HashMap::new();
            $(
                _map.insert($key, $value);
            )*
            _map 
        }
    }
}
