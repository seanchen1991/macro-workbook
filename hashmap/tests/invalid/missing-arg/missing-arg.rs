use hashmap::hashmap;
use std::collections::HashMap;

fn main() {
    // an arg should be present between each pair of commas 
    let map: HashMap<_, _> = hashmap!{1 => 'a', , 2 => 'b'};
}
