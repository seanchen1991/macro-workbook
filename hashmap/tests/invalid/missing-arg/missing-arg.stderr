error: no rules expected the token `,`
 --> $DIR/missing-arg.rs:6:49
  |
6 |     let map: HashMap<_, _> = hashmap!{1 => 'a', , 2 => 'b'};
  |                                                 ^ no rules expected this token in macro call
